orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/ManifestForceCacheRefreshTask.java

```diff
@@ -17,48 +17,250 @@
+        if (processedManifests.contains(id)) {
+          continue;
+        }
+
+        Optional<PendingRefresh> pendingRefresh = pendingRefreshes.stream()
+            .filter(pr -> pr.getDetails() != null)
+            .filter(pr -> account.equals(pr.getDetails().getAccount()) &&
+                location.equals(pr.getDetails().getLocation()) &&
+                name.equals(pr.getDetails().getName())
+            )
+            .findAny();
+
+        if (pendingRefresh.isPresent()) {
+          if (!pendingRefreshProcessed(pendingRefresh.get(), refreshedManifests, startTime)) {
+            allProcessed = false;
+          }
+        } else {
+          log.warn("No pending refresh of {} in {}", id, account);
+          allProcessed = false;
+          refreshedManifests.remove(id);
+        }
+      }
+    }
+
+    return new TaskResult(allProcessed ? SUCCEEDED : RUNNING, toContext(stageData));
+  }
+
+  private boolean pendingRefreshProcessed(PendingRefresh pendingRefresh, Set<String> refreshedManifests, long startTime) {
+    PendingRefresh.Details details = pendingRefresh.getDetails();
+    if (pendingRefresh.cacheTime == null || pendingRefresh.processedTime == null || details == null) {
+      log.warn("Pending refresh of {} is missing cache metadata", pendingRefresh);
+      refreshedManifests.remove(toManifestIdentifier(details.getLocation(), details.getName()));
+      return false;
+    } else if (pendingRefresh.cacheTime < startTime) {
+      log.warn("Pending refresh of {} is stale", pendingRefresh);
+      refreshedManifests.remove(toManifestIdentifier(details.getLocation(), details.getName()));
+      return false;
+    } else if (pendingRefresh.processedTime < startTime) {
+      log.info("Pending refresh of {} was cached as a part of this request, but not processed", pendingRefresh);
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  private Map<String, List<String>> manifestsNeedingRefresh(StageData stageData) {
+    Map<String, List<String>> deployedManifests = stageData.getManifestNamesByNamespace();
+    Set<String> refreshedManifests = stageData.getRefreshedManifests();
+    if (deployedManifests.isEmpty()) {
+      log.warn("No manifests were deployed, nothing to refresh...");
+    }
+
+    Map<String, List<String>> result = new HashMap<>();
+    for (Map.Entry<String, List<String>> entry : deployedManifests.entrySet()) {
+      String location = entry.getKey();
+      List<String> names = entry.getValue().stream()
+          .filter(n -> !refreshedManifests.contains(toManifestIdentifier(location, n)))
+          .collect(Collectors.toList());
+
+      if (!names.isEmpty()) {
+        result.put(location, names);
+      }
+    }
+
+    return result;
+  }
+
+  private boolean refreshManifests(String provider, String account, StageData stageData) {
+    Map<String, List<String>> manifests = manifestsNeedingRefresh(stageData);
+
+    final boolean[] allRefreshesSucceeded = {true};
```

Comment:
Is this so you can set a variable from within a lambda?

Reply:
Yup, `#java`